### PR TITLE
[DeviceSanitizer] Support detecting out-of-bounds errors on sycl::buffer

### PR DESCRIPTION
**Pull request by zhaomaosu**
As we discussed before, we intercepted buffer with usm for short-term solution. For long-term, we will draft a new ocl extension to allow user to create buffer with fixed address.

SYCLOS PR for pre-ci: https://github.com/intel/llvm/pull/13504

### CODE REVIEW

File: source/loader/layers/sanitizer/asan_buffer.cpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/asan_buffer.cpp:123]: The `64` here seems not match the `128` in LINE 132 below.

**Review by AllanZyne — APPROVED** [review]

**Comment by zhaomaosu** [comment]
@oneapi-src/unified-runtime-maintain, could you please help review this PR when you are available? Thanks.

File: source/loader/layers/sanitizer/ur_sanddi.cpp
Comment (pbalcer) [inline source/loader/layers/sanitizer/ur_sanddi.cpp:1081]: This is not really thread-safe. If multiple `urKernelRelease` are done in parallel, multiple threads may read RefCount == 1. I think you might need to have your own RefCount like for MemBuffer.

File: source/loader/layers/sanitizer/asan_buffer.cpp
Comment (pbalcer) [inline source/loader/layers/sanitizer/asan_buffer.cpp:73]: Wouldn't it be easier to just store the pointer instead of Parent ptr and an offset?

File: source/loader/layers/sanitizer/ur_sanddi.cpp
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/ur_sanddi.cpp]: Thanks, you are right. Updated.

File: source/loader/layers/sanitizer/asan_buffer.cpp
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_buffer.cpp:73]: We want buffer/subbuffer do the allocation when they are actually used. If we store the pointer, this means we need to do the allocation when subbuffer is created. This is not we expected.

**Review by pbalcer — APPROVED** [review]

**Comment by zhaomaosu** [comment]
@oneapi-src/unified-runtime-maintain, any other comments for this PR? Thanks.

**Comment by kbenzie** [comment]
> @oneapi-src/unified-runtime-maintain, any other comments for this PR? Thanks.

Don't think so. If you could make https://github.com/intel/llvm/pull/13504 ready for review so to get approvals in place there before we merge here that would be great.

**Pull request merged**
